src/outputs/rcp.c

```diff
@@ -0,0 +1,1484 @@
+
+  // s is now the official device session
+  outputs_device_session_add(device->id, s);
+
+  DPRINTF(E_INFO, L_RCP, "Connection to '%s' established%s\n", s->devname, s->force_reboot_on_fatal ? " (will force reboot on fatal)" : "");
+
+  return s;
+
+ out_free_ev:
+  event_free(s->reply_timeout);
+  event_free(s->ev);
+ out_close_connection:
+  rcp_disconnect(s->sock);
+ out_free_session:
+  free(s);
+
+  return NULL;
+}
+
+
+static void
+rcp_session_free(struct rcp_session *s)
+{
+  if (!s)
+    return;
+
+  DPRINTF(E_DBG, L_RCP, "session free %ld %s at %s\n", s->device->id, s->devname, s->address);
+
+
+  if (s->sock >= 0)
+    rcp_disconnect(s->sock);
+  if (s->ev)
+    event_free(s->ev);
+  if (s->reply_timeout)
+    event_free(s->reply_timeout);
+
+  free(s->stream_url);
+  free(s->address);
+  free(s->devname);
+
+  free(s);
+}
+
+static void
+rcp_session_cleanup(struct rcp_session *rs)
+{
+  struct rcp_session *s;
+
+  DPRINTF(E_DBG, L_RCP, "session cleanup %ld %s at %s\n", rs->device->id, rs->devname, rs->address);
+
+  if (rs == rcp_sessions)
+    rcp_sessions = rcp_sessions->next;
+  else
+    {
+      for (s = rcp_sessions; s && (s->next != rs); s = s->next)
+	; /* EMPTY */
+
+      if (!s)
+	DPRINTF(E_WARN, L_RCP, "WARNING: struct rcp_session not found in list; BUG!\n");
+      else
+	s->next = rs->next;
+    }
+
+  outputs_device_session_remove(rs->device->id);
+
+  rcp_session_free(rs);
+}
+
+
+/* ---------------------------- STATUS HANDLERS ----------------------------- */
+
+static void
+rcp_status(struct rcp_session *s)
+{
+  enum output_device_state state;
+
+  switch (s->state)
+    {
+      case RCP_STATE_SETUP:
+	state = OUTPUT_STATE_STARTUP;
+	break;
+
+      case RCP_STATE_QUEUING_CLEAR ... RCP_STATE_QUEUING_PLAY:
+      case RCP_STATE_VOL_GET:
+      case RCP_STATE_VOL_SET:
+      case RCP_STATE_STOPPING ... RCP_STATE_SHUTDOWN_SERVER_DISCONNECT:  // as per @frankusb, move advise that remote/we're not interested
+	state = OUTPUT_STATE_CONNECTED;
+	break;
+
+      case RCP_STATE_STREAMING:
+	state = OUTPUT_STATE_STREAMING;
+	break;
+
+      case RCP_STATE_STOPPED ... RCP_STATE_DISCONNECTED:
+	state = OUTPUT_STATE_STOPPED;
+	break;
+
+      case RCP_STATE_FAILED:
+      default:
+	state = OUTPUT_STATE_FAILED;
+    }
+
+  DPRINTF(E_DBG, L_RCP, "state -> %d\n", state);
+  outputs_cb(s->callback_id, s->device->id, state);
+  s->callback_id = -1;
+
+  if (state == OUTPUT_STATE_STOPPED || state == OUTPUT_STATE_FAILED)
+    rcp_session_cleanup(s);
+}
+
+
+/* ------------------ INTERFACE FUNCTIONS CALLED BY OUTPUTS.C --------------- */
+
+static int
+rcp_device_start(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s;
+
+  DPRINTF(E_DBG, L_RCP, "device start %ld '%s'\n", device->id, device->name);
+
+  s = rcp_session_make(device, callback_id);
+  if (!s)
+    return -1;
+
+  rcp_status(s);
```

Comment:
Normally the status is sent async? I.e. after the device has replied with some sort of ok.